crates/lib/src/execution_data.rs

```diff
@@ -1,120 +1,345 @@
-                    .map(|(program_name, _)| program_name.as_ref())
-                    .map(parse_readout_register)
-                    .map(|reference| Ok((reference?, values.clone())))
-                    .collect::<Result<Vec<_>, MemoryReferenceParseError>>()
-            })
-            .flatten()
-            .collect::<HashMap<_, _>>()
-            .into();
-
-        result
-    }
-}
+                    // Pair all the memory references with their readout values
+                    .map(|(memory_reference, alias)| {
+                        Ok((
+                            parse_readout_register(memory_reference).map_err(|e| {
+                                RegisterMatrixConversionError::MemoryReferenceParseError(e)
+                            })?,
+                            readout_data.readout_values.get(alias).ok_or(
+                                RegisterMatrixConversionError::UnmappedAlias {
+                                    memory_reference: memory_reference.to_string(),
+                                    alias: alias.to_string(),
+                                },
+                            )?,
+                        ))
+                    })
+                    // Collect into a type that will sort them by memory reference, this allows us
+                    // to make sure indices are sequential.
+                    .collect::<Result<
+                        BTreeMap<MemoryReference, &ReadoutValues>,
+                        RegisterMatrixConversionError,
+                    >>()?
+                    .iter()
+                    // Iterate over them in reverse. Starting with the last index lets us:
+                    //     (1): Initialize the RegisterMatrix with the correct number of rows
+                    //     (2): Use a memory reference with an index of 0 as a setinel.
+                    .rev()
+                    .try_fold(
+                        (
+                            HashMap::with_capacity(readout_data.readout_values.len()),
+                            None::<&MemoryReference>,
+                        ),
+                        |(mut readout_data, previous_reference), (reference, values)| {
+                            // If we haven't started on a new register, make sure we aren't
+                            // skipping an index. For example, if we jumped from ro[5] to ro[3], or
+                            // ro[2] to theta[1], we skipped a column.
+                            if let Some(previous) = previous_reference {
+                                if previous.name != reference.name
+                                    || previous.index != reference.index + 1
+                                {
+                                    return Err(RegisterMatrixConversionError::MissingRow {
+                                        register: reference.name.clone(),
+                                        index: previous.index + 1,
```

Comment:
The condition is `previous.index != reference.index + 1` but the returned index is `previous.index + 1`. Looks like a potential off by one error to me.

Also, would it be worth doing this check for all registers before trying to build the matrix? You use a `BTreeMap` to sort by key, but collecting into `Vec` would also allow you to sort, and then `Vec::windows(2)` would allow you to iterate over sequential pairs of registers and make this check before doing any of the below work, including any cloning it may do.

May be overthinking this, but the logic here is somewhat complex and I'm looking for ways to make it easier to follow. Maybe intermediate values would be enough to make it easier to follow.

Reply:
Right, fixed the condition by using `previous.index - 1` in both places instead.

Reply:
Yeah, the logic here is complex. I initially wanted to a check like you suggest with `windows` and `any`, but stopped short when I realized you couldn't use the former with iterators. 

I can try collecting it into an intermediate value, but wasn't sure if that was worth breaking the chain.

Reply:
I used itertools to break the sequence check into it's own step.